### PR TITLE
CodeGen: Don't assert when printing null GlobalAddress operands

### PR DESCRIPTION
**Pull request by arsenm**

### CODE REVIEW

**Comment by arsenm** [comment]
* **#115531** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/115531?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by qcolombet — APPROVED** [review]
I'm guessing this is for mir files.
I ran into this problem in the past.

LGTM, but could you add a mir test and here add a comment that specifically says that this shouldn't be the state of a well formed ir.  (Unless you're fixing a real bug, and then we are missing a test anyway)

**Comment by qcolombet** [comment]
> > I'm guessing this is for mir files. I ran into this problem in the past.
> > LGTM, but could you add a mir test and here add a comment that specifically says that this shouldn't be the state of a well formed ir. (Unless you're fixing a real bug, and then we are missing a test anyway)
> 
> I don't think this can be constructed from MIR. A literal 0 won't parse as a GlobalAddress

So how do you run into this problem?

It sounds to me that we shouldn't produce nullptr for GlobalAddress in a normal compilation flow.

**Comment by arsenm** [comment]
> So how do you run into this problem?

Bugs

> It sounds to me that we shouldn't produce nullptr for GlobalAddress in a normal compilation flow.

Yes, but this is in debug printing. -debug causing new crashes to appear is a bad property

**Comment by qcolombet** [comment]
> > So how do you run into this problem?
> 
> Bugs
> 
> > It sounds to me that we shouldn't produce nullptr for GlobalAddress in a normal compilation flow.
> 
> Yes, but this is in debug printing. -debug causing new crashes to appear is a bad property

Fair enough.
Could you add a comment that says that this is for debug printing?

File: llvm/lib/CodeGen/MachineOperand.cpp [hunk 1]
```diff
@@ -909,7 +909,11 @@ void MachineOperand::print(raw_ostream &OS, ModuleSlotTracker &MST,
     OS << printJumpTableEntryReference(getIndex());
     break;
   case MachineOperand::MO_GlobalAddress:
-    getGlobal()->printAsOperand(OS, /*PrintType=*/false, MST);
+    if (const auto *GV = getGlobal())
+      getGlobal()->printAsOperand(OS, /*PrintType=*/false, MST);
+    else
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineOperand.cpp:914]:
```suggestion
    else // Invalid, but may appear in debugging scenarios.
```

**Pull request merged**